core/src/util/url.cpp

```diff
@@ -0,0 +1,487 @@
+    return path;
+}
+
+void Url::parse() {
+
+    // The parsing process roughly follows https://tools.ietf.org/html/rfc1808#section-2.4
+
+    size_t start = 0;
+    size_t end = buffer.size();
+
+    // Parse the fragment.
+    {
+        // If there's a '#' in the string, the substring after it to the end is the fragment.
+        auto pound = std::min(buffer.find('#', start), end);
+        parts.fragment.start = std::min(pound + 1, end);
+        parts.fragment.count = end - parts.fragment.start;
+
+        // Remove the '#' and fragment from parsing.
+        end = pound;
+    }
+
+    // Parse the scheme.
+    {
+        size_t i = start;
+        auto c = buffer[i];
+
+        // A scheme is permitted to contain only alphanumeric characters, '+', '.', and '-'.
+        while (i < end && (isalnum(c) || c == '+' || c == '.' || c == '-')) {
+            c = buffer[++i];
+        }
+
+        // If a scheme is present, it must be followed by a ':'.
+        if (c == ':') {
+            parts.scheme.start = start;
+            parts.scheme.count = i - start;
+
+            // Remove the scheme and ':' from parsing.
+            start = i + 1;
+
+            // Set the absolute URL flag.
+            flags |= IS_ABSOLUTE;
+        }
+    }
+
+    // If scheme is 'data', parse as data URI.
+    if (buffer.compare(parts.scheme.start, parts.scheme.count, "data") == 0) {
+
+        // Set the data scheme flag.
+        flags |= HAS_DATA_SCHEME;
+
+        // A data scheme will be followed by a media type, then either a comma or a base 64 indicator string.
+        auto base64Indicator = std::min(buffer.find(";base64", start), end);
+        auto comma = std::min(buffer.find(',', start), end);
+
+        // If the base 64 indicator string is found before the comma, set the matching flag.
+        if (base64Indicator < comma) {
+            flags |= HAS_BASE64_DATA;
+        }
+
+        // The media type goes from the colon after the scheme up to either the comma or the base 64 string.
+        parts.media.start = start;
+        parts.media.count = std::min(base64Indicator, comma) - start;
+
+        // The data section is separated by a comma and goes to the end of the URI.
+        start = std::min(comma + 1, end);
+        parts.data.start = start;
+        parts.data.count = end - start;
+
+        // We're done!
+        return;
+    }
+
+    // Check whether the scheme is 'http', 'https', or 'file' and set appropriate flags.
+    if (buffer.compare(parts.scheme.start, parts.scheme.count, "http") == 0 ||
+        buffer.compare(parts.scheme.start, parts.scheme.count, "https") == 0) {
+        flags |= HAS_HTTP_SCHEME;
+    } else if (buffer.compare(parts.scheme.start, parts.scheme.count, "file") == 0) {
+        flags |= HAS_FILE_SCHEME;
+    }
+
+    // If '//' is next in the string, then the substring up to the following '/' is the network location.
+    if (buffer.compare(start, 2, "//") == 0) {
+        start += 2;
```

Comment:
This should only be checked when a scheme was found and it's not a `FILE_SCHEME`